### PR TITLE
Require return type for functions (#94)

### PR DESCRIPTION
**Pull request by dmikis**
For #94

### CODE REVIEW

**Comment by ikokostya** [comment]
I think in some situations omitting return type is legal.

**Example 1** Generating user data

```ts
const obj = {
    a: 1,
    b: 'str'
};

sendJson(obj).catch((err) => console.error(err));
```
where `sendJson()` function has the following signature

```ts
declare function sendJson(obj: any): Promise<void>;
```
In the example above we don't declare `obj` as `any` and allow infer its type to TypeScript. We also don't define new type, because it's useless (`obj` only passed to `sendJson()` function, which takes `any` as argument).

After that add some dynamic value to the `obj`:

```ts
function createObj(a: number) {
    return {
        a: a,
        b: 'str'
    };
}

sendJson(createObj(1)).catch((err) => console.error(err));
```
Type of the `createObj` function also is not specified, because it's useless here and adds additional noise for reader. I think it's better then `any` return type, because `any` completely disable type checks. Now we just say: *I don't interesting about return type*. 

`sendJson()` can be replaced to `Response#send()` function of `express` or `Client#query()` function of `node-postgres` library.

In other words: *return type can be ommited if result will be passed to function which takes more wide type*.

**Example 2** Interface implementations.

We have some interface:

```ts
type Position = [number, number];
type BBox = [Position, Position];

interface Geometry {
    getBBox(): Bbox;
    getCenter(): Position;
}
```
and implementation for many different geometries: `Point`, `MultiPoint`, `Rectangle`, `LineString`, `Polygon`, `Circle`, ...

Return type in each implementation may be safely omitted for simplicity. TypeScript shows error if return type violates interface.

**Comment by dmikis** [comment]
I believe both examples not only won't suffer from specifying return types, but may actually benefit from it.

If you don't care about type then cast it to `{}`. That will prevent you (or somebody else, which's more likely) from relying on properties of the infered type. But if you need to rely on any structural information about the type, then the whole "don't care" premise is planly wrong: you should describe that type.

In the second case is somewhat more viable, however I don't see providing return type to a method overload that big of a hustle to introduce an exception to the rule.

**Comment by ikokostya** [comment]
> If you don't care about type then cast it to {}.

Type `{}` allow any object, but return type of the function is narrow: it has concrete properties. For example, for variable declaration type inference is allowed:

```ts
const obj: {} = {a: 1}; // Bad
const obj = {a: 1}; // Good
```

> whole "don't care" premise is planly wrong

It's wrong if type used often in other places. But for single use it's ok and better than `any` or `{}`, which looses type checks.

Always specifying complete type sounds good in theory, but in practice it's waste of time in provided cases.

**Comment by dmikis** [comment]
>  But for single use it's ok and better than any or {}, which looses type checks

How?

Also, I suggest we wait on more input on the issue.

**Comment by ikokostya** [comment]
> How?

```ts
function f(a: {a: number}) {}

function g(): {} {
    return  {a: 1};
}

f(g()); // Argument of type '{}' is not assignable to parameter of type '{ a: number; }'.
          // Property 'a' is missing in type '{}'.
```

**Comment by dmikis** [comment]
You breaking away of your own logic behind the example: if you pass return value of a function into another one, you should correctly specify it's type if definition of the function.

The reasoning is that the rule should be simple. "Types everywhere" is simple. If we can devise a simple enough exception, I'm all for it.

**Comment by ikokostya** [comment]
> if you pass return value of a function into another one, you should correctly specify it's type if definition of the function.

For variables we don't specify type

```ts
const obj = {a: 1};
f(obj);
```

For function I suggest the same approach: 

```ts
function createObj(a: number) {
    return {a: a};
}

f(createObj(1));
```

**Comment by dmikis** [comment]
> Also, anonymous functions are not described here: for anonymous functions type inference is allowed?

That's actually a good point. Usually for arrow function passed as callback we don't even specify types of parameters. I would like to incorporate that into style guide.

**Comment by ikokostya** [comment]
@dmikis Ok, let's add note for anonymous functions and merge this.

File: typescript.md [hunk 1]
```diff
@@ -105,6 +105,43 @@ from JS).
...
+
+  **Bad**
+
+  ```ts
+  function foo() {
+      // ...
+  }
+
+  function bar() {
+      return something ? 1 | undefined;
```
Comment (ikokostya) [inline typescript.md]: `| -> :`

**Comment by ikokostya** [comment]
Also, we can add check this rule with `tslint`: https://palantir.github.io/tslint/rules/typedef/

```json
"typedef": [
  true,
  "call-signature"
]
```
